---
title: "Random Generation of Gene IDs and SNP Rate Calculation"
author: "Elena Venable"
date: "February 5, 2015"
output: html_document
---

## 1. Load the Necessary Data
The output of the `categorize_SNP` function is all_compiled_data, which can be used to calculate SNP rates.  Also load a file containing all the gene_IDs and the gene_lists for which a snp rate is needed.

```{r}
library(SexSpecificExpression)
load("~/SexSpecificExpression/data/gene_IDs.rda")
load("~/SexSpecificExpression/data/all_compiled_data.rda")
load("~/SexSpecificExpression/data/gene_lists_three_sets.rda")
```

```{r,include=FALSE}
load("~/SexSpecificExpression/data/snp_rate.rda")
```
Calculating SNP rates requires the compiled genome data, which is a subset of the `all_compiled_data` list. Below, the genome is extracted from the data.

```{r}
genome <- all_compiled_data$genome_with_SNP
```

## 2. Run the Background Rate Simulation and Calculation Function
Use the code below to run the code that simulates the background rate and calculates the rate of SNPs in different gene lists. The variable `rep_num` is the amount of simulations that will be used to estimate the background rate.  The output of the function has all the trials of the simulation and a table containing the true and simulated rates.  The table output is shown below

```{r,eval=FALSE}
snp_rate <- compare_background_rate(gene_list = gene_lists_three_sets,all_IDs = gene_IDs,genome = genome,rep_num = 100)
```
```{r}
snp_rate$rates
```